chi/server.py

```diff
@@ -798,15 +800,36 @@ def associate_floating_ip(server_id, floating_ip_address=None):
         floating_ip_address (str): The IPv4 address of the Floating IP to
             assign. If specified, this Floating IP must already be allocated
             to the project.
+        port_id (str): Optional port ID to assign the floating IP to. If not
+            provided, the will use the first routable port on the server.
 
     """
     if not floating_ip_address:
-        floating_ip_address = get_free_floating_ip()["floating_ip_address"]
-
-    conn = connection()
-    conn.add_ips_to_server(conn.get_server_by_id(server_id), ips=[floating_ip_address])
+        floating_ip_obj = chi_network.get_free_floating_ip()
+    else:
+        floating_ip_obj = chi_network.get_floating_ip(floating_ip_address)
 
-    return floating_ip_address
+    conn = connection(session=session())
+    if port_id:
+        try:
+            return conn.network.update_ip(
+                floating_ip_obj["id"], port_id=port_id,
+            )["floating_ip_address"]
+        except SDKException:
+            raise exception.ResourceError(f"Port {port_id} cannot be assigned floating ip {floating_ip_address} on server {server_id}")
```

Comment:
I would do this after listing the ports attached to the server, and ensure that this port is in the list of ports attached to the server, so we don't let the user accidentally bind an address to some other port.